ouroboros-network/src/Ouroboros/Network/Diffusion.hs

```diff
@@ -1,1320 +1,406 @@
-                -> (LocalSnocket -> LocalSocket -> IO a)
-                -> IO a
-withLocalSocket iocp tracer localAddress k =
-  bracket
-    (
-      case localAddress of
-#if defined(mingw32_HOST_OS)
-         -- Windows uses named pipes so can't take advantage of existing sockets
-         Left _ -> traceWith tracer UnsupportedReadySocketCase
-                >> throwIO UnsupportedReadySocket
-#else
-         Left sd -> do
-             addr <- Socket.getSocketName sd
-             case addr of
-                  (Socket.SockAddrUnix path) -> do
-                    traceWith tracer (UsingSystemdSocket path)
-                    return (Left ( Snocket.localSnocket iocp path
-                                 , (LocalSocket sd)
-                                 ))
-                  _  -> do
-                    traceWith tracer $ UnsupportedLocalSystemdSocket addr
-                    throwIO UnsupportedLocalSocketType
-#endif
-         Right addr -> do
-             let sn :: LocalSnocket
-                 sn = Snocket.localSnocket iocp addr
-             traceWith tracer $ CreateSystemdSocketForSnocketPath addr
-             sd <- Snocket.open sn Snocket.LocalFamily
-             traceWith tracer $ CreatedLocalSocket addr
-             return (Right (sn, sd, addr))
-    )
-    -- We close the socket here, even if it was provided to us.
-    (\case
-      Left  (sn, sd)    -> Snocket.close sn sd
-      Right (sn, sd, _) -> Snocket.close sn sd)
-    $ \case
-      -- unconfigured socket
-      Right (sn, sd, addr) -> do
-        traceWith tracer . ConfiguringLocalSocket addr
-           =<< localSocketFileDescriptor sd
-        Snocket.bind sn sd (NodeToClient.LocalAddress addr)
-        traceWith tracer . ListeningLocalSocket addr
-           =<< localSocketFileDescriptor sd
-        Snocket.listen sn sd
-        traceWith tracer . LocalSocketUp addr
-           =<< localSocketFileDescriptor sd
-        k sn sd
-
-      -- pre-configured systemd socket
-      Left (sn, sd) -> k sn sd
+  :: DiffusionTracers
+  -> DiffusionArguments IO
+  -> DiffusionApplications
+       RemoteAddress
+       LocalAddress
+       NodeToNodeVersionData
+       NodeToClientVersionData
+       IO
+  -> IO ()
+runDataDiffusion
+  (DiffusionTracers
+    tr@Common.DiffusionTracers
+       { dtP2P })
+  (DiffusionArguments
+    dargs@Common.DiffusionArguments
+       { daP2P })
+  (DiffusionApplications
+    dapps@Common.DiffusionApplications
+       { dapP2P }) =
+  case (dtP2P, daP2P, dapP2P) of
+    (Left t, Left da, Left dapp)    ->
+      NonP2P.runDataDiffusion
+        (tr { dtP2P = t})
+        (dargs { daP2P = da })
+        (dapps { dapP2P = dapp })
+    (Right t, Right da, Right dapp) ->
+      void
+        $ P2P.runDataDiffusion
+            (tr { dtP2P = t})
+            (dargs { daP2P = da })
+            (dapps { dapP2P = dapp })
+    _                               ->
+      error "Non-matching arguments, every argument should be on the same side!"
```

Comment:
We could eliminate it with a type level tag
```
data P2PMode = P2P | NonP2P
```
And then index all types with either `P2P` or `NonP2P`.  But maybe it is not worth it, I'll leave it up to. 

Hydra guys are also using data diffusion interface as well, we might need to help them to make a transition.

Reply:
I don't think its worth it, the smart constructors will take care of everything and if someone uses the wrong smart constructors, I think the compiler can not type check